### PR TITLE
Fix Narrator cannot announce the progress of the VScrollbar/HScrollba…

### LINKED ISSUE #9158: [Accessibility] Narrator cannot announce the progress of the VScrollbar/HScrollbar when updating it
### .NET version

.Net 8.0 SDK build: 8.0.100-preview.5.23268.13

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No

### Issue description

When using keyboard to update the progress of the VScrollbar/HScrollbar controls, Narrator cannot announce the progress updating

.Net 8.0:
image
ScrollbarProgress
image

### Steps to reproduce

Repro steps:
1. Create a Winforms .net core application with VScrollbar/HScrollbar
2. Build and run application
3. Launch Narrator, and use keyboard to focus on VScrollbar/HScrollbar control to update its progress 

or use this sample application:
[WinFormsApp19.zip](https://github.com/dotnet/winforms/files/11529647/WinFormsApp19.zip)

### PR DESCRIPTION
**Pull request by SimonZhao888**
Fixes #9158 
RangeValue property determines whether Scroll Bar value  should be read out by screen reader as the Scroll Bar Range Value changed.

## Proposed changes

- Implement RangeValue pattern in ScrollBarAccessibleObject.
- Raising RangeValueValuePropertyId property-changed event when scroll bar value is changed.

## Customer Impact

- Narrator cannot announce the progress of the VScrollbar/HScrollbar when updating it.

## Regression? 

- No

## Risk

-Minimal

### Before
RangeValue Patten is not displayed in the Insights properties panel.
image
image

### After
RangeValue Patten is displayed in the Insights properties panel.
image
ScrollBar
Narrator Buddy output as range value is changed into the Scroll Bar:

1,
2,
3,
4,
5,
6,
7,

## Test methodology 

- Manual testing
- Unit tests
- CTI

## Accessibility testing  

- Using Narrator and AI

 

## Test environment(s) 

- dotnet 8.0.0-preview.6.23280.5
- Windows 11

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9217)

### CODE REVIEW

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks great! Please add description of testing done to the PR template and a test case. When opening a draft PR, please describe what the next steps are in the PR description.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 1]
```diff
@@ -141,18 +141,43 @@ UiaCore.UIA.ControlTypePropertyId when
...
+
+        internal override double SmallChange => double.NaN;
+
+        internal override double Maximum => _owningScrollBar.Maximum;
+
+        internal override double Minimum => _owningScrollBar.Minimum;
+
+        internal override bool IsReadOnly => true;
+
         internal override bool IsPatternSupported(UiaCore.UIA patternId)
```
Comment (v-elnovikova) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:169]: Small nit:
```suggestion
        internal override bool IsPatternSupported(UiaCore.UIA patternId)
            => patternId switch
            {
                UiaCore.UIA.ValuePatternId => true,
                UiaCore.UIA.RangeValuePatternId => true,
                _ => base.IsPatternSupported(patternId)
            };
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 2]
```diff
@@ -141,18 +141,43 @@ UiaCore.UIA.ControlTypePropertyId when
...
+                UiaCore.UIA.IsValuePatternAvailablePropertyId => true,
+                UiaCore.UIA.IsRangeValuePatternAvailablePropertyId => true,
                 _ => base.GetPropertyValue(propertyID)
             };
 
         internal override bool IsIAccessibleExSupported() => true;
 
+        internal override double RangeValue => _owningScrollBar.Value;
+
+        internal override double LargeChange => double.NaN;
```
Comment (v-elnovikova) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:159]: Shouldn't this property also return value from `_owningScrollBar`, like `Minimum` and `Maximum`?

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 3]
```diff
@@ -141,18 +141,43 @@ UiaCore.UIA.ControlTypePropertyId when
...
+                UiaCore.UIA.RangeValueMaximumPropertyId => Maximum,
+                UiaCore.UIA.RangeValueMinimumPropertyId =>Minimum,
+                UiaCore.UIA.IsValuePatternAvailablePropertyId => true,
+                UiaCore.UIA.IsRangeValuePatternAvailablePropertyId => true,
                 _ => base.GetPropertyValue(propertyID)
             };
 
         internal override bool IsIAccessibleExSupported() => true;
 
+        internal override double RangeValue => _owningScrollBar.Value;
```
Comment (v-elnovikova) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:157]: Could you please check if RangeValue pattern expects percentage here (`0...1.0` or maybe `0...100`), not an absolute value?

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:157]: Following is the RangeValue pattern Screenshot for Winforms, WPF, and OutLook:
Winforms:
image

WPF:
image

Outlook:
image

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs (hunk 1 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:169]: Done.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs (hunk 2 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs]: Done.

**Comment by Tanya-Solyanik** [comment]
Looks good! Please move the PR from `Draft` state after addressing the comments. Test failures seem to be unrelated.

**Review by dmitrii-drobotov — COMMENTED** [review]
Could you also add screenshots with affected property values in Inspect or Accessibility Insights before and after the changes to the PR description? And also Narrator output when pressing arrow keys on a scroll bar, so that the impact is more clear. As an example, you can take a look at description in https://github.com/dotnet/winforms/pull/7825 or https://github.com/dotnet/winforms/pull/8329.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObjectTests.cs [hunk 4]
```diff
@@ -51,6 +51,11 @@ public void ScrollBarAccessibleObject_IsPatternSupported_Invoke_ReturnsExpected(
     [InlineData((int)UIA.IsKeyboardFocusablePropertyId, true)]
     [InlineData((int)UIA.IsValuePatternAvailablePropertyId, true)]
     [InlineData((int)UIA.AutomationIdPropertyId, "AutomId")]
+    [InlineData((double)UIA.RangeValueMaximumPropertyId, (double)100)]
+    [InlineData((double)UIA.RangeValueMinimumPropertyId, (double)0)]
+    [InlineData((double)UIA.RangeValueValuePropertyId, (double)0)]
+    [InlineData((double)UIA.RangeValueLargeChangePropertyId, (double)10)]
```
[the remark this answers is not in the document]
Comment (dkazennov) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObjectTests.cs:57]: Good point here.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObjectTests.cs [hunk 5]
```diff
@@ -139,11 +144,13 @@ public void ScrollBarAccessibleObject_GetPropertyValue_IsEnabled_ReturnsExpected
...
     [InlineData(false, ((int)UIA.IsTogglePatternAvailablePropertyId))]
     [InlineData(true, ((int)UIA.IsValuePatternAvailablePropertyId))]
+    [InlineData(true, ((int)UIA.RangeValueIsReadOnlyPropertyId))]
+    [InlineData(true, ((int)UIA.IsRangeValuePatternAvailablePropertyId))]
     public void ScrollBarAccessibleObject_GetPropertyValue_Pattern_ReturnsExpected(bool expected, int propertyId)
     {
         using SubScrollBar scrollBar = new() { Enabled = true };
         ScrollBar.ScrollBarAccessibleObject accessibleObject = (ScrollBar.ScrollBarAccessibleObject)scrollBar.AccessibilityObject;
-
+
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObjectTests.cs:153]: Please remove the extra spaces.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 6]
```diff
@@ -141,19 +142,47 @@ UiaCore.UIA.ControlTypePropertyId when
...
-            if (patternId == UiaCore.UIA.ValuePatternId)
+            if (!_owningScrollBar.IsHandleCreated)
             {
-                return true;
+                return;
             }
 
-            return base.IsPatternSupported(patternId);
+            _owningScrollBar.Value = (int)newValue;
+            base.SetValue(newValue);
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:177]: I think this line can be removed because the base method is empty.
```suggestion
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 7]
```diff
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Drawing;
+
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:6]:
```suggestion
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObjectTests.cs (hunk 5 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObjectTests.cs]: Done.

**Review by dmitrii-drobotov — COMMENTED** [review]
Looks good! Could you also add a unit test for SetValue method? I think you can take test values from here:
https://github.com/dotnet/winforms/blob/67d9b72c120283fb44428e0add323c8a5c119bc7/src/System.Windows.Forms/tests/InteropTests/AccessibleObjectTests.cs#L145-L151
and then assert that calling it changes `scrollBar.Value`. For out of range values you would probably need a separate test which asserts that correct exception is thrown.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs (hunk 7 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs]: Done.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs (hunk 6 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs]: Done.

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs]: I hade added two unit test.
One is used to test the SetValue method, and another one is to test the exception message for out of range values.

**Review by dmitrii-drobotov — COMMENTED** [review]
Just a couple more points regarding tests, otherwise looks good!

**Review by Tanya-Solyanik — APPROVED** [review]
